bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/VeluxBridgeHandler.java

```diff
@@ -253,71 +257,88 @@ public boolean bridgeCommunicate(BridgeCommunicationProtocol communication) {
-        logger.info("Initializing Velux Bridge '{}'.", getThing().getUID());
-        // The framework requires you to return from this method quickly.
-        // Setting the thing status to UNKNOWN temporarily and let the background task decide for the real status.
-        logger.trace("initialize() called.");
+        // set the thing status to UNKNOWN temporarily and let the background task decide the real status
         updateStatus(ThingStatus.UNKNOWN);
-        // Take care of unusual situations...
+        // take care of unusual situations...
         if (scheduler.isShutdown()) {
-            logger.warn("initialize(): scheduler is shutdown, aborting the initialization of this bridge.");
+            logger.warn("initialize(): scheduler is shutdown, aborting initialization.");
             return;
         }
-        getTaskExecutor();
-        logger.trace("initialize(): preparing background initialization task.");
-        // Background initialization...
         scheduler.execute(() -> {
-            logger.trace("initialize.scheduled(): Further work within scheduler.execute().");
-            logger.trace("initialize.scheduled(): Initializing bridge configuration parameters.");
-            this.veluxBridgeConfiguration = new VeluxBinding(getConfigAs(VeluxBridgeConfiguration.class)).checked();
-            logger.trace("initialize.scheduled(): work on updated bridge configuration parameters.");
-            bridgeParamsUpdated();
-
-            logger.debug("initialize.scheduled(): activated scheduler with {} milliseconds.",
-                    this.veluxBridgeConfiguration.refreshMSecs);
-            refreshJob = scheduler.scheduleWithFixedDelay(() -> {
-                try {
-                    refreshOpenHAB();
-                } catch (RuntimeException e) {
-                    logger.warn("Exception occurred during activated refresh scheduler: {}.", e.getMessage());
-                }
-            }, this.veluxBridgeConfiguration.refreshMSecs, this.veluxBridgeConfiguration.refreshMSecs,
-                    TimeUnit.MILLISECONDS);
-            logger.trace("initialize.scheduled(): done.");
+            initializeScheduled();
         });
-        logger.trace("initialize() done.");
     }
 
     /**
-     * NOTE: It takes care about shutting down the connections before removal of this binding.
+     * Various initialisation actions to be executed on a background thread
+     *
+     * Note: if a disposeScheduled() task has been scheduled for the same ThingUID, then execution of this method will
+     * be paused until that task has finished.
      */
+    private synchronized void initializeScheduled() {
+        logger.trace("initializeScheduled(): initialize bridge configuration parameters.");
+        veluxBridgeConfiguration = new VeluxBinding(getConfigAs(VeluxBridgeConfiguration.class)).checked();
+        Future<Boolean> disposeTask = disposeTasks.remove(veluxBridgeConfiguration.ipAddress);
```

Comment:
Can follow this approach !

But it seems to be a workaround of misbehaviour of the framework, isn't it?

Reply:
> a workaround of misbehaviour of the framework

I suppose you are right. The framework certainly should not call initialize() and dispose() in parallel on the same Thing, but there are two issues that would hinder their serialization..
- The coding guidelines ask that initialize() and dispose() must return quickly; which is is why in our case initialize() and dispose() submit respective xxxScheduled() tasks to the scheduler, and return themselves fast. So although the framework could serialize initialize() and dispose(), it couldn't serialize the respective xxxScheduled() tasks.
- The need to serialize initialize() and dispose() methods specifically only applies to a special case where both initialize() and dispose() would be working on the unique same hub (in our case with the same IP address) aka 'Representation Property'. But framework has no way to know about that.

So I am going to ping @kaikreuzer to see if he has any opinion on this?

Reply:
> The framework certainly should not call initialize() and dispose() in parallel

This is in fact guaranteed, so there is no need to synchronize the methods.

> So although the framework could serialize initialize() and dispose(), it couldn't serialize the respective xxxScheduled() tasks.

That's correct. Any long running tasks that are scheduled as asynchronous tasks must be correctly handled by the binding.
Afaik, you don't have to expect initialize() to be called after dispose(), though. The only situation you must be able to deal with is a dispose() call while some job that was started in initialize() might still be running. But the usual logic is to simply cancel/interrupt such tasks.

Reply:
@kaikreuzer many thanks for the helpful information.

> you don't have to expect initialize() to be called after dispose()

Unfortunately in the case of the Velux, we do have to expect (nay ensure) just that: The Velux bridge is notorious for getting into a "zombie" state if it is not shutdown properly. It only allows one socket to be opened on it at a time, and will reject additional socket connect requests so long as a prior socket is still connected. And so if the OH binding just "goes away" the bridge may still believe that there is a socket connected (even though there is no activity). And the bridge seems not to use keep-alives or time outs. So it is essential that the OH binding dispose() method is always called, and always fully completed (powering Off the OH computer is not a good thing). This means the OH binding must allow any already started comms tasks to complete, flush any remaining data from the bridges send buffer, and then do a proper FIN-ACK-FIN-ACK shutdown on the socket. I am pretty sure that the changes in this PR now ensure just that (excepting the PC power Off situation of course..)